cdx.go

```diff
@@ -0,0 +1,270 @@
+	CDXCanonizedHost
+	CDXCanonizedImage
+	CDXCanonizedJumpPoint
+	CDXUnknownFBISChanges // 'K'
+	CDXCanonizedLink
+	CDXMetaTags
+	CDXMassagedURL
+	_ // 'O'
+	CDXCanonizedPath
+	CDXLanguage
+	CDXCanonizedRedirect
+	CDXCompressedSize
+	_ // 'T'
+	CDXUniqness
+	CDXCompressedOffset
+	_ // 'W'
+	CDXCanonizedHrefURL
+	CDXCanonizedSrcURL
+	CDXCanonizedScriptURL // 'Z'
+)
+const (
+	CDXOriginalURL = 'a' + iota
+	CDXDate
+	CDXOldChecksum
+	CDXUncompressedDATOffset
+	CDXIP
+	CDXFrame
+	CDXArcFileName
+	CDXOriginalHost
+	CDXImage
+	CDXJumpPoint
+	CDXDigest
+	CDXLink
+	CDXMimeType
+	CDXUncompressedSize
+	CDXPort
+	CDXOriginalPath
+	_ // 'q'
+	CDXRedirect
+	CDXResponseCode
+	CDXTitle
+	CDXUUID // 'u'
+	CDXUncompressedOffset
+	_ // 'w'
+	CDXHrefURL
+	CDXSrcURL
+	CDXScriptURL // 'z'
+
+	CDXComment = '#'
+)
+
+func iaMassageHost(host string) string {
+	rgx := regexp.MustCompile(`www\d*\.`)
+	m := rgx.FindStringIndex(host)
+	if m != nil {
+		return host[m[1]:]
+	}
+	return host
+}
+
+func surtHost(host string) string {
+	// ip addresses ARE reversed
+	split := strings.Split(host, ".")
+	for i, j := 0, len(split)-1; i < j; i, j = i+1, j-1 {
+		split[i], split[j] = split[j], split[i]
+	}
+	return strings.Join(split, ",")
+}
+
+func alphaReorderQuery(query string) string {
+	if len(query) <= 1 {
+		return query
+	}
+	split := strings.Split(query, "&")
+	// this is a deviation from the python version
+	// I can't tell if the split on = actually does anything useful
+	sort.Strings(split)
+	return strings.Join(split, "&")
+}
+
+func iaMassagedURL(u1 *url.URL) string {
+	u := new(url.URL)
+	*u = *u1
+	u.Host = strings.ToLower(u.Host)
+	host, port, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		host = u.Host
+	} else if err == nil {
+		if u.Scheme == "http" && port == "80" {
+			port = ""
+		} else if u.Scheme == "https" && port == "443" {
+			port = ""
+		}
+	}
+	host = iaMassageHost(host)
+	u.Scheme = ""
+	u.User = nil
+	u.Path = strings.ToLower(u.Path)
+	// u.Path = stripPathSessionID(u.Path)
+	if true { // path_strip_trailing_slash_unless_empty
+		if u.Path != "/" {
+			u.Path = strings.TrimSuffix(u.Path, "/")
+		}
+	}
+	if u.RawQuery != "" {
+		// u.RawQuery = stripQuerySessionID(u.RawQuery)
+		u.RawQuery = strings.ToLower(u.RawQuery)
+		u.RawQuery = alphaReorderQuery(u.RawQuery)
+	}
+	u.ForceQuery = false
+	u.Fragment = ""
+	// -----
+	host = surtHost(host)
+	if port != "" {
+		u.Host = host + ":" + port + ")"
+	} else {
+		u.Host = host + ")"
+	}
+	u.Scheme = "XXX"
+	return strings.TrimPrefix(u.String(), "XXX://")
+}
+
+// The CDXFormat type maps a CDX header character (the key) to an array index.
+// Values should be contiguous.
+type CDXFormat map[byte]int
+
+// Writes the CDX fields that can be determined from the record into the target
+// array.  Not all fields are implemented or can be implemented, see source for
+// details.  Fields not written are left at their original values.
+func (r *Record) CDXLine(format CDXFormat, line []string) error {
```

Comment:
cdx.go:151:1: comment on exported method Record.CDXLine should be of the form "CDXLine ..."